File: csharp/Assembler/Loader/Normal/LoaderNormal.cs
using Arshu.App.Json;
using Arshu.Common;
using Assembler.Common;
using Assembler.Interface;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;

namespace Assembler.Loader.Normal;

/// <summary>
/// Loader implementation for EngineNormal
/// Handles loading and caching of HTML and JSON templates from the file system
/// Templates are kept immutable within the loader
/// </summary>
public class LoaderNormal : ILoaderNormal
{
    private static readonly ConcurrentDictionary<string, Dictionary<string, (string html, string? json)>> _htmlTemplatesCache = new();
    private Dictionary<string, (string html, string? json)> _templates;
    private Dictionary<string, string> _parentMap;
    private string _appSite;

    #region Constructor

    /// <summary>
    /// Creates a new LoaderNormal instance (use Load() method to load templates)
    /// </summary>
    public LoaderNormal()
    {
        _templates = new Dictionary<string, (string html, string? json)>();
        _parentMap = new Dictionary<string, string>();
        SearchAppSites = string.Empty;
        _appSite = string.Empty;
    }

    /// <summary>
    /// Convenience constructor that automatically loads templates
    /// </summary>
    /// <param name="rootDirPath">Root directory path containing AppSites folder</param>
    /// <param name="appSite">Primary AppSite name to load</param>
    /// <param name="searchAppSites">Comma-delimited string of AppSite names for fallback templates</param>
    public LoaderNormal(string rootDirPath, string appSite, string searchAppSites) : this()
    {
        Load(rootDirPath, appSite, searchAppSites);
    }

    #endregion

    #region ILoaderNormal Interface

    #region Loading Related

    /// <summary>
    /// Gets the search AppSites for template fallback resolution
    /// </summary>
    public string SearchAppSites { get; private set; }

    /// <summary>
    /// Loads all templates from the specified directory
    /// Returns true if loading succeeded, false otherwise
    /// </summary>
    public bool Load(string rootDirPath, string appSite, string searchAppSites)
    {
        try
        {
            Logger.Debug($"Load called for appSite: {appSite}, searchAppSites: {searchAppSites}", "LoaderNormal");

            _appSite = appSite;
            SearchAppSites = searchAppSites;

            var cacheKey = Path.GetDirectoryName(rootDirPath) + "|" + appSite + "|" + searchAppSites;
            if (_htmlTemplatesCache.TryGetValue(cacheKey, out var cached))
            {
                Logger.Debug($"Returning cached templates for {appSite} ({cached.Count} templates)", "LoaderNormal");
                _templates = cached;
                return true;
            }

            // Load templates from primary appSite
            _templates = LoadTemplatesFromSingleAppSite(rootDirPath, appSite);

            // Load templates from searchAppSites for fallback
            if (!string.IsNullOrEmpty(searchAppSites))
            {
                var searchAppSitesArray = searchAppSites.Split(',');
                for (int i = 0; i < searchAppSitesArray.Length; i++)
                {
                    var searchAppSite = searchAppSitesArray[i].Trim();
                    if (string.IsNullOrEmpty(searchAppSite))
                        continue;

                    var searchTemplates = LoadTemplatesFromSingleAppSite(rootDirPath, searchAppSite);
                    foreach (var kvp in searchTemplates)
                    {
                        // Only add if not already present (primary appSite takes precedence)
                        if (!_templates.ContainsKey(kvp.Key))
                        {
                            _templates[kvp.Key] = kvp.Value;
                            Logger.Debug($"Added fallback template '{kvp.Key}' from '{searchAppSite}'", "LoaderNormal");
                        }
                    }
                }
            }

            _htmlTemplatesCache.TryAdd(cacheKey, _templates);

            // Build parent-child relationship map for JSON inheritance
            _parentMap = BuildParentMap();
            Logger.Debug($"Built parent map with {_parentMap.Count} relationships for JSON inheritance", "LoaderNormal");

            Logger.Info($"Loaded {_templates.Count} templates for {appSite}", "LoaderNormal");
            return true;
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to load templates: {ex.Message}", "LoaderNormal");
            return false;
        }
    }

    /// <summary>
    /// Checks if a template exists
    /// </summary>
    public bool HasTemplate(string appSite, string templateName)
    {
        var key = $"{appSite.ToLowerInvariant()}_{templateName.ToLowerInvariant()}";
        return _templates.ContainsKey(key);
    }

    /// <summary>
    /// Clears the template cache
    /// </summary>
    public void ClearCache()
    {
        _htmlTemplatesCache.Clear();
    }

    #endregion

    #region Merging Related

    /// <summary>
    /// Gets a template's HTML content by appSite and name with AppView fallback support
    /// </summary>
    public string? GetTemplateHtml(string appSite, string templateName, string? appView = null, string? appViewPrefix = null)
    {
        // Try AppView fallback first if provided
        if (!string.IsNullOrEmpty(appView) && !string.IsNullOrEmpty(appViewPrefix) &&
            templateName.Contains(appViewPrefix, StringComparison.OrdinalIgnoreCase))
        {
            var appKey = CommonUtil.ReplaceCaseInsensitive(templateName, appViewPrefix, appView);
            var fallbackKey = $"{appSite.ToLowerInvariant()}_{appKey.ToLowerInvariant()}";

            if (_templates.TryGetValue(fallbackKey, out var appViewTemplate))
            {
                return appViewTemplate.html;
            }
        }

        // Try primary template key
        var key = $"{appSite.ToLowerInvariant()}_{templateName.ToLowerInvariant()}";
        if (_templates.TryGetValue(key, out var template))
        {
            return template.html;
        }

        // Try searchAppSites fallback
        if (!string.IsNullOrEmpty(SearchAppSites))
        {
            var searchAppSitesArray = SearchAppSites.Split(',');
            for (int i = 0; i < searchAppSitesArray.Length; i++)
            {
                var searchAppSite = searchAppSitesArray[i].Trim();
                if (string.IsNullOrEmpty(searchAppSite))
                    continue;

                var searchKey = $"{searchAppSite.ToLowerInvariant()}_{templateName.ToLowerInvariant()}";
                if (_templates.TryGetValue(searchKey, out var fallbackTemplate))
                {
                    Logger.Debug($"Template '{templateName}' not found in '{appSite}', using fallback from '{searchAppSite}'", "LoaderNormal");
                    return fallbackTemplate.html;
                }
            }
        }

        return null;
    }

    /// <summary>
    /// Merges HTML string with JSON data for the specified template
    /// This centralizes JSON merging logic in the loader for clean architecture
    /// The loader internally retrieves JSON (with inheritance resolution) and merges it with HTML
    /// Engines cannot access raw JSON - all JSON retrieval and merging happens inside the loader
    /// </summary>
    public string MergeHtmlWithJson(string html, string appSite, string templateName)
    {
        Logger.Debug($"MergeHtmlWithJson called: appSite={appSite}, templateName={templateName}", "LoaderNormal");

        if (string.IsNullOrEmpty(html))
            return html;

        // Get JSON with inheritance resolution
        var jsonData = GetTemplateJsonWithInheritance(appSite, templateName);

        if (jsonData == null)
        {
            Logger.Debug($"No JSON data found for {templateName}, returning original HTML", "LoaderNormal");
            return html;
        }

        var jsonKeys = string.Join(", ", jsonData.Keys);
        Logger.Debug($"Merging HTML with JSON for {templateName} (keys: {jsonKeys})", "LoaderNormal");
        return JsonMergeUtil.MergeTemplateWithJson(html, jsonData);
    }

    #endregion

    #endregion

    #region Loading Templates (Private)

    /// <summary>
    /// Loads templates from a single AppSite without caching or fallback logic
    /// </summary>
    private static Dictionary<string, (string html, string? json)> LoadTemplatesFromSingleAppSite(string rootDirPath, string appSite)
    {
        var result = new Dictionary<string, (string html, string? json)>();
        var appSitesPath = Path.Combine(rootDirPath, "AppSites", appSite);

        if (!Directory.Exists(appSitesPath))
        {
            Logger.Warn($"AppSites directory not found: {appSitesPath}", "LoaderNormal");
            return result;
        }

        Logger.Debug($"Loading templates from: {appSitesPath}", "LoaderNormal");

        foreach (var file in Directory.GetFiles(appSitesPath, "*.html", SearchOption.AllDirectories))
        {
            var fileName = Path.GetFileNameWithoutExtension(file);
            var key = ($"{appSite.ToLowerInvariant()}_{fileName.ToLowerInvariant()}");
            var htmlContent = CommonUtil.NormalizeFileContent(File.ReadAllText(file));

            Logger.Debug($"Loading template: {key} (html size: {htmlContent.Length})", "LoaderNormal");

            // Find JSON file case-insensitively
            var jsonFile = Path.ChangeExtension(file, ".json");
            string? jsonContent = null;

            // Try exact match first
            if (File.Exists(jsonFile))
            {
                jsonContent = CommonUtil.NormalizeFileContent(File.ReadAllText(jsonFile));
                Logger.Debug($"Found JSON file for {key} (size: {jsonContent.Length})", "LoaderNormal");
            }
            else
            {
                // Try case-insensitive search in the same directory
                var directory = Path.GetDirectoryName(file);
                var baseFileName = Path.GetFileNameWithoutExtension(file);
                if (!string.IsNullOrEmpty(directory))
                {
                    var jsonFiles = Directory.GetFiles(directory, "*.json", SearchOption.TopDirectoryOnly);
                    string? matchingJson = null;
                    for (int i = 0; i < jsonFiles.Length; i++)
                    {
                        if (string.Equals(Path.GetFileNameWithoutExtension(jsonFiles[i]), baseFileName, StringComparison.OrdinalIgnoreCase))
                        {
                            matchingJson = jsonFiles[i];
                            break;
                        }
                    }

                    if (matchingJson != null)
                    {
                        jsonContent = CommonUtil.NormalizeFileContent(File.ReadAllText(matchingJson));
                        Logger.Debug($"Found JSON file (case-insensitive) for {key} (size: {jsonContent.Length})", "LoaderNormal");
                    }
                }
            }

            result[key] = (htmlContent, jsonContent);
        }

        Logger.Debug($"Loaded {result.Count} templates for {appSite}", "LoaderNormal");
        return result;
    }

    #endregion

    #region JSON Inheritance Support (Private)

    // NOTE: The following methods are INTENTIONAL COPIES used across all loaders for architectural separation.
    // Each loader/engine pair is independent to allow individual evolution without shared dependencies.
    // DO NOT extract these to shared utilities - that would create tight coupling.

    /// <summary>
    /// Builds a parent-child relationship map by analyzing template placeholders
    /// Tracks which template is the parent of another based on {{TemplateName}} references
    /// </summary>
    private Dictionary<string, string> BuildParentMap()
    {
        var parentMap = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        Logger.Debug($"Building parent map for appSite: {_appSite}", "LoaderNormal");

        foreach (var kvp in _templates)
        {
            var templateKey = kvp.Key;
            var html = kvp.Value.html;

            // Find all {{TemplateName}} placeholders in this template
            var searchPos = 0;
            while (searchPos < html.Length)
            {
                var openStart = html.IndexOf("{{", searchPos);
                if (openStart == -1) break;

                // Skip special placeholders (#, @, $, /)
                if (openStart + 2 < html.Length &&
                    (html[openStart + 2] == '#' || html[openStart + 2] == '@' ||
                     html[openStart + 2] == '$' || html[openStart + 2] == '/'))
                {
                    searchPos = openStart + 2;
                    continue;
                }

                var closeStart = html.IndexOf("}}", openStart + 2);
                if (closeStart == -1) break;

                var placeholderName = html.Substring(openStart + 2, closeStart - openStart - 2).Trim();

                // Check if this is a valid alphanumeric template name
                if (!string.IsNullOrEmpty(placeholderName) && IsAlphaNumeric(placeholderName))
                {
                    // This template (templateKey) is the parent of the placeholder template
                    var childTemplateKey = $"{_appSite.ToLowerInvariant()}_{placeholderName.ToLowerInvariant()}";

                    if (!parentMap.ContainsKey(childTemplateKey))
                    {
                        parentMap[childTemplateKey] = templateKey;
                        Logger.Debug($"Parent relationship: {childTemplateKey} -> parent: {templateKey}", "LoaderNormal");
                    }
                }

                searchPos = closeStart + 2;
            }
        }

        Logger.Debug($"Built parent map with {parentMap.Count} relationships", "LoaderNormal");
        return parentMap;
    }

    /// <summary>
    /// Gets parsed JSON with inheritance resolution
    /// Resolves keys ending with # by searching up the parent tree
    /// </summary>
    private JsonObject? GetTemplateJsonWithInheritance(string appSite, string templateName)
    {
        var templateKey = $"{appSite.ToLowerInvariant()}_{templateName.ToLowerInvariant()}";
        Logger.Debug($"GetTemplateJsonWithInheritance: templateKey={templateKey}", "LoaderNormal");

        // Try to get JSON from primary appSite template
        string? jsonContent = null;
        if (_templates.TryGetValue(templateKey, out var template))
        {
            jsonContent = template.json;
        }
        else
        {
            // Try searchAppSites fallback
            if (!string.IsNullOrEmpty(SearchAppSites))
            {
                var searchAppSitesArray = SearchAppSites.Split(',');
                for (int i = 0; i < searchAppSitesArray.Length; i++)
                {
                    var searchAppSite = searchAppSitesArray[i].Trim();
                    if (string.IsNullOrEmpty(searchAppSite))
                        continue;

                    var searchKey = $"{searchAppSite.ToLowerInvariant()}_{templateName.ToLowerInvariant()}";
                    if (_templates.TryGetValue(searchKey, out var fallbackTemplate))
                    {
                        jsonContent = fallbackTemplate.json;
                        templateKey = searchKey; // Update key for inheritance resolution
                        break;
                    }
                }
            }
        }

        if (string.IsNullOrEmpty(jsonContent))
        {
            Logger.Debug($"No JSON found for templateKey={templateKey}", "LoaderNormal");
            return null;
        }

        // Parse JSON string
        JsonObject jsonObj;
        try
        {
            jsonObj = JsonConverter.ParseJsonString(jsonContent);
        }
        catch (Exception ex)
        {
            Logger.Error($"Failed to parse JSON for {templateKey}: {ex.Message}", "LoaderNormal");
            return null;
        }

        var rawKeys = string.Join(", ", jsonObj.Keys);
        Logger.Debug($"Raw JSON keys for {templateKey}: {rawKeys}", "LoaderNormal");

        var resolvedJson = new JsonObject();

        // Process each JSON key and resolve inheritance
        foreach (var kvp in jsonObj)
        {
            var key = kvp.Key;
            var value = kvp.Value;

            // Check if this is an inheritable key (ends with #)
            if (key.EndsWith("#") && value is string strValue)
            {
                // Resolve inherited value
                var actualKey = key.Substring(0, key.Length - 1);
                Logger.Debug($"Found inheritance key: {key}, defaultValue={strValue}, resolving for actualKey={actualKey}", "LoaderNormal");
                var resolvedValue = ResolveJsonKeyWithInheritance(actualKey, strValue, templateKey);
                if (resolvedValue != null)
                {
                    resolvedJson[actualKey] = resolvedValue;
                    Logger.Debug($"Resolved inherited key {key} -> {actualKey} = {resolvedValue}", "LoaderNormal");
                    continue;
                }
            }

            // Normal key - keep as is
            resolvedJson[key] = value;
        }

        return resolvedJson;
    }

    /// <summary>
    /// Resolves a JSON key by searching up the parent tree
    /// </summary>
    private string? ResolveJsonKeyWithInheritance(string actualKey, string defaultValue, string currentTemplateKey)
    {
        Logger.Debug($"Resolving inherited key: {actualKey} for template {currentTemplateKey}", "LoaderNormal");

        // Search up the parent tree for the key
        var inheritedValue = SearchParentTreeForKey(actualKey, currentTemplateKey);

        if (inheritedValue != null)
        {
            Logger.Debug($"Found inherited value for {actualKey}: {inheritedValue}", "LoaderNormal");
            return inheritedValue;
        }

        // If not found in parents, use the default value
        Logger.Debug($"No inherited value found for {actualKey}, using default: {defaultValue}", "LoaderNormal");
        return defaultValue;
    }

    /// <summary>
    /// Searches up the parent tree to find a JSON key value
    /// </summary>
    private string? SearchParentTreeForKey(string key, string currentTemplateKey)
    {
        // Get parent template key
        if (!_parentMap.TryGetValue(currentTemplateKey, out var parentKey))
        {
            Logger.Debug($"No parent found for {currentTemplateKey}", "LoaderNormal");
            return null;
        }

        Logger.Debug($"Checking parent {parentKey} for key {key}", "LoaderNormal");

        // Get parent's template
        if (!_templates.TryGetValue(parentKey, out var parentTemplate))
        {
            Logger.Debug($"Parent template {parentKey} not found in templates", "LoaderNormal");
            return null;
        }

        if (string.IsNullOrEmpty(parentTemplate.json))
        {
            Logger.Debug($"Parent template {parentKey} has no JSON data, searching further up", "LoaderNormal");
            // Parent has no JSON, search further up the tree
            return SearchParentTreeForKey(key, parentKey);
        }

        // Parse parent's JSON
        try
        {
            var parentJsonObj = JsonConverter.ParseJsonString(parentTemplate.json);

            // Look for the key (case-insensitive)
            foreach (var kvp in parentJsonObj)
            {
                if (string.Equals(kvp.Key, key, StringComparison.OrdinalIgnoreCase))
                {
                    if (kvp.Value is string strValue)
                    {
                        Logger.Debug($"Found key {key} in parent {parentKey}: {strValue}", "LoaderNormal");
                        return strValue;
                    }
                }
            }

            Logger.Debug($"Key {key} not found in parent {parentKey}, searching further up", "LoaderNormal");
            // Not found in this parent, search further up the tree
            return SearchParentTreeForKey(key, parentKey);
        }
        catch (Exception ex)
        {
            Logger.Error($"Error parsing JSON for parent {parentKey}: {ex.Message}", "LoaderNormal");
            return null;
        }
    }

    private static bool IsAlphaNumeric(string str)
    {
        for (int i = 0; i < str.Length; i++)
        {
            if (!char.IsLetterOrDigit(str[i]))
                return false;
        }
        return true;
    }

    #endregion
}

File: node/Assembler/src/loader/normal/loaderNormal.js
// Normal loader equivalent to C# ILoaderNormal
import fs from 'fs';
import path from 'path';
import { Logger } from '../../../../Arshu/src/common/Logger.js';
import { normalizeFileContent, replaceCaseInsensitive } from '../../common/commonUtil.js';
import { mergeTemplateWithJson } from '../jsonMergeUtil.js';

export class LoaderNormal {
    static #cache = new Map(); // key: root|appSite|searchAppSites -> Map(key -> {html,json})
    #rootDirPath;
    #appSite;
    #templates;
    #parentMap; // For JSON inheritance

    constructor(rootDirPath, appSite, searchAppSites = '') {
        this.#rootDirPath = rootDirPath || '';
        this.#appSite = appSite || '';
        this.searchAppSites = searchAppSites || '';
        this.#templates = new Map();
        this.#parentMap = new Map();

        // Auto-load if parameters are provided (matching C# pattern)
        if (rootDirPath && appSite !== undefined) {
            this.load(rootDirPath, appSite, searchAppSites);
        }
    }

    // ILoaderNormal: load
    load(rootDirPath, appSite, searchAppSites) {
        if (rootDirPath) this.#rootDirPath = rootDirPath;
        if (appSite) this.#appSite = appSite;
        if (typeof searchAppSites === 'string') this.searchAppSites = searchAppSites;

        const cacheKey = `${path.dirname(this.#rootDirPath)}|${this.#appSite}|${this.searchAppSites}`;
        if (LoaderNormal.#cache.has(cacheKey)) {
            this.#templates = LoaderNormal.#cache.get(cacheKey);
            Logger.debug(`LoaderNormal: using cached templates for ${this.#appSite} (${this.#templates.size})`, 'LoaderNormal');
            return true;
        }

        const result = new Map();
        // primary
        this.#loadTemplatesFromSingleAppSite(this.#rootDirPath, this.#appSite, result);
        // fallbacks
        if (this.searchAppSites) {
            for (const raw of this.searchAppSites.split(',')) {
                const site = raw.trim();
                if (!site) continue;
                this.#loadTemplatesFromSingleAppSite(this.#rootDirPath, site, result, true);
            }
        }
        this.#templates = result;
        LoaderNormal.#cache.set(cacheKey, result);

        // Build parent-child relationship map for JSON inheritance
        this.#parentMap = this.BuildParentMap();
        Logger.debug(`Built parent map with ${this.#parentMap.size} relationships for JSON inheritance`, 'LoaderNormal');

        return true;
    }

    // ILoaderNormal: hasTemplate
    hasTemplate(appSite, templateName) {
        const key = `${appSite.toLowerCase()}_${templateName.toLowerCase()}`;
        return this.#templates.has(key);
    }

    // ILoaderNormal: clearCache
    static clearCache() {
        LoaderNormal.#cache.clear();
    }

    // Instance clearCache for interface compatibility
    clearCache() {
        LoaderNormal.#cache.clear();
    }

    // ILoaderNormal: getTemplateHtml with AppView + SearchAppSites logic
    getTemplateHtml(appSite, templateName, appView = null, appViewPrefix = null) {
        const rec = this.#getTemplateInternal(appSite, templateName, appView, appViewPrefix);
        return rec ? rec.html : null;
    }

    // ILoaderNormal: mergeHtmlWithJson centralizing JSON merging
    mergeHtmlWithJson(html, appSite, templateName) {
        if (!html) return html;

        // Get JSON with inheritance resolution (matching C# GetTemplateJsonWithInheritance)
        const jsonObj = this.GetTemplateJsonWithInheritance(appSite, templateName);
        if (!jsonObj) return html;

        return mergeTemplateWithJson(html, jsonObj);
    }

    // ============== internals ==============
    #getTemplateInternal(appSite, templateName, appView, appViewPrefix) {
        const viewPrefix = appViewPrefix || '';
        if (appView && viewPrefix && templateName.toLowerCase().includes(viewPrefix.toLowerCase())) {
            const appKey = replaceCaseInsensitive(templateName, viewPrefix, appView);
            const fallbackKey = `${appSite.toLowerCase()}_${appKey.toLowerCase()}`;
            if (this.#templates.has(fallbackKey)) return this.#templates.get(fallbackKey);
        }
        const primaryKey = `${appSite.toLowerCase()}_${templateName.toLowerCase()}`;
        if (this.#templates.has(primaryKey)) return this.#templates.get(primaryKey);
        if (this.searchAppSites) {
            for (const raw of this.searchAppSites.split(',')) {
                const site = raw.trim();
                if (!site) continue;
                const searchKey = `${site.toLowerCase()}_${templateName.toLowerCase()}`;
                if (this.#templates.has(searchKey)) {
                    Logger.debug(`LoaderNormal: fallback template '${templateName}' from '${site}'`, 'LoaderNormal');
                    return this.#templates.get(searchKey);
                }
            }
        }
        return null;
    }

    #loadTemplatesFromSingleAppSite(rootDirPath, appSite, result, isFallback = false) {
        const appSitesPath = path.join(rootDirPath, 'AppSites', appSite);
        if (!fs.existsSync(appSitesPath) || !fs.statSync(appSitesPath).isDirectory()) {
            Logger.warn(`AppSites directory not found: ${appSitesPath}`, 'LoaderNormal');
            return;
        }
        Logger.debug(`LoaderNormal: loading ${isFallback ? 'fallback ' : ''}templates from ${appSitesPath}`, 'LoaderNormal');
        this.#walkDirectory(appSitesPath, (filePath, stats) => {
            if (!stats.isFile() || path.extname(filePath).toLowerCase() !== '.html') return;
            const fileName = path.basename(filePath, '.html');
            const key = `${appSite.toLowerCase()}_${fileName.toLowerCase()}`;
            if (result.has(key)) return; // primary wins
            const htmlContent = normalizeFileContent(fs.readFileSync(filePath, 'utf8'));
            let jsonContent = null;
            const jsonFile = filePath.replace(/\.html$/i, '.json');
            if (fs.existsSync(jsonFile)) {
                jsonContent = normalizeFileContent(fs.readFileSync(jsonFile, 'utf8'));
            } else {
                const dir = path.dirname(filePath);
                const baseName = path.basename(filePath, '.html').toLowerCase();
                const entries = fs.readdirSync(dir);
                for (const entry of entries) {
                    const entryPath = path.join(dir, entry);
                    if (fs.statSync(entryPath).isFile() && path.extname(entry).toLowerCase() === '.json') {
                        const entryBase = path.basename(entry, path.extname(entry)).toLowerCase();
                        if (entryBase === baseName) {
                            jsonContent = normalizeFileContent(fs.readFileSync(entryPath, 'utf8'));
                            break;
                        }
                    }
                }
            }
            result.set(key, { html: htmlContent, json: jsonContent });
        });
    }

    #walkDirectory(dir, callback) {
        const files = fs.readdirSync(dir);
        for (const file of files) {
            const filePath = path.join(dir, file);
            const stats = fs.statSync(filePath);
            callback(filePath, stats);
            if (stats.isDirectory()) this.#walkDirectory(filePath, callback);
        }
    }

    // ============== JSON Inheritance Support (Private) ==============
    // NOTE: The following methods are INTENTIONAL COPIES used across all loaders for architectural separation.
    // Each loader/engine pair is independent to allow individual evolution without shared dependencies.
    // DO NOT extract these to shared utilities - that would create tight coupling.

    /**
     * Builds a parent-child relationship map by analyzing template placeholders
     * Tracks which template is the parent of another based on {{TemplateName}} references
     * Matches C# BuildParentMap()
     */
    BuildParentMap() {
        const parentMap = new Map();

        Logger.debug(`Building parent map for appSite: ${this.#appSite}`, 'LoaderNormal');

        for (const [templateKey, template] of this.#templates) {
            const html = template.html;

            // Find all {{TemplateName}} placeholders in this template
            let searchPos = 0;
            while (searchPos < html.length) {
                const openStart = html.indexOf('{{', searchPos);
                if (openStart === -1) break;

                // Skip special placeholders (#, @, $, /)
                if (openStart + 2 < html.length &&
                    (html[openStart + 2] === '#' || html[openStart + 2] === '@' ||
                     html[openStart + 2] === '$' || html[openStart + 2] === '/')) {
                    searchPos = openStart + 2;
                    continue;
                }

                const closeStart = html.indexOf('}}', openStart + 2);
                if (closeStart === -1) break;

                const placeholderName = html.substring(openStart + 2, closeStart).trim();

                // Check if this is a valid alphanumeric template name
                if (placeholderName && this.IsAlphaNumeric(placeholderName)) {
                    // This template (templateKey) is the parent of the placeholder template
                    const childTemplateKey = `${this.#appSite.toLowerCase()}_${placeholderName.toLowerCase()}`;

                    if (!parentMap.has(childTemplateKey)) {
                        parentMap.set(childTemplateKey, templateKey);
                        Logger.debug(`Parent relationship: ${childTemplateKey} -> parent: ${templateKey}`, 'LoaderNormal');
                    }
                }

                searchPos = closeStart + 2;
            }
        }

        Logger.debug(`Built parent map with ${parentMap.size} relationships`, 'LoaderNormal');
        return parentMap;
    }

    /**
     * Gets parsed JSON with inheritance resolution
     * Resolves keys ending with # by searching up the parent tree
     * Matches C# GetTemplateJsonWithInheritance()
     */
    GetTemplateJsonWithInheritance(appSite, templateName) {
        const templateKey = `${appSite.toLowerCase()}_${templateName.toLowerCase()}`;
        Logger.debug(`GetTemplateJsonWithInheritance: templateKey=${templateKey}`, 'LoaderNormal');

        // Try to get JSON from primary appSite template
        let jsonContent = null;
        let actualTemplateKey = templateKey;

        if (this.#templates.has(templateKey)) {
            jsonContent = this.#templates.get(templateKey).json;
        } else {
            // Try searchAppSites fallback
            if (this.searchAppSites) {
                for (const searchAppSite of this.searchAppSites.split(',')) {
                    const site = searchAppSite.trim();
                    if (!site) continue;

                    const searchKey = `${site.toLowerCase()}_${templateName.toLowerCase()}`;
                    if (this.#templates.has(searchKey)) {
                        jsonContent = this.#templates.get(searchKey).json;
                        actualTemplateKey = searchKey; // Update key for inheritance resolution
                        break;
                    }
                }
            }
        }

        if (!jsonContent) {
            Logger.debug(`No JSON found for templateKey=${templateKey}`, 'LoaderNormal');
            return null;
        }

        // Parse JSON string
        let jsonObj;
        try {
            jsonObj = JSON.parse(jsonContent);
        } catch (ex) {
            Logger.error(`Failed to parse JSON for ${templateKey}: ${ex.message}`, 'LoaderNormal');
            return null;
        }

        const rawKeys = Object.keys(jsonObj).join(', ');
        Logger.debug(`Raw JSON keys for ${templateKey}: ${rawKeys}`, 'LoaderNormal');

        const resolvedJson = {};

        // Process each JSON key and resolve inheritance
        for (const [key, value] of Object.entries(jsonObj)) {
            // Check if this is an inheritable key (ends with #)
            if (key.endsWith('#') && typeof value === 'string') {
                // Resolve inherited value
                const actualKey = key.substring(0, key.length - 1);
                Logger.debug(`Found inheritance key: ${key}, defaultValue=${value}, resolving for actualKey=${actualKey}`, 'LoaderNormal');
                const resolvedValue = this.ResolveJsonKeyWithInheritance(actualKey, value, actualTemplateKey);
                if (resolvedValue !== null) {
                    resolvedJson[actualKey] = resolvedValue;
                    Logger.debug(`Resolved inherited key ${key} -> ${actualKey} = ${resolvedValue}`, 'LoaderNormal');
                    continue;
                }
            }

            // Normal key - keep as is
            resolvedJson[key] = value;
        }

        return resolvedJson;
    }

    /**
     * Resolves a JSON key by searching up the parent tree
     * Matches C# ResolveJsonKeyWithInheritance()
     */
    ResolveJsonKeyWithInheritance(actualKey, defaultValue, currentTemplateKey) {
        Logger.debug(`Resolving inherited key: ${actualKey} for template ${currentTemplateKey}`, 'LoaderNormal');

        // Search up the parent tree for the key
        const inheritedValue = this.SearchParentTreeForKey(actualKey, currentTemplateKey);

        if (inheritedValue !== null) {
            Logger.debug(`Found inherited value for ${actualKey}: ${inheritedValue}`, 'LoaderNormal');
            return inheritedValue;
        }

        // If not found in parents, use the default value
        Logger.debug(`No inherited value found for ${actualKey}, using default: ${defaultValue}`, 'LoaderNormal');
        return defaultValue;
    }

    /**
     * Searches up the parent tree to find a JSON key value
     * Matches C# SearchParentTreeForKey()
     */
    SearchParentTreeForKey(key, currentTemplateKey) {
        // Get parent template key
        if (!this.#parentMap.has(currentTemplateKey)) {
            Logger.debug(`No parent found for ${currentTemplateKey}`, 'LoaderNormal');
            return null;
        }

        const parentKey = this.#parentMap.get(currentTemplateKey);
        Logger.debug(`Checking parent ${parentKey} for key ${key}`, 'LoaderNormal');

        // Get parent's template
        if (!this.#templates.has(parentKey)) {
            Logger.debug(`Parent template ${parentKey} not found in templates`, 'LoaderNormal');
            return null;
        }

        const parentTemplate = this.#templates.get(parentKey);
        if (!parentTemplate.json) {
            Logger.debug(`Parent template ${parentKey} has no JSON data, searching further up`, 'LoaderNormal');
            // Parent has no JSON, search further up the tree
            return this.SearchParentTreeForKey(key, parentKey);
        }

        // Parse parent's JSON
        try {
            const parentJsonObj = JSON.parse(parentTemplate.json);

            // Look for the key (case-insensitive)
            for (const [jsonKey, jsonValue] of Object.entries(parentJsonObj)) {
                if (jsonKey.toLowerCase() === key.toLowerCase()) {
                    if (typeof jsonValue === 'string') {
                        Logger.debug(`Found key ${key} in parent ${parentKey}: ${jsonValue}`, 'LoaderNormal');
                        return jsonValue;
                    }
                }
            }

            Logger.debug(`Key ${key} not found in parent ${parentKey}, searching further up`, 'LoaderNormal');
            // Not found in this parent, search further up the tree
            return this.SearchParentTreeForKey(key, parentKey);
        } catch (ex) {
            Logger.error(`Error parsing JSON for parent ${parentKey}: ${ex.message}`, 'LoaderNormal');
            return null;
        }
    }

    /**
     * Checks if a string contains only alphanumeric characters
     * Matches C# IsAlphaNumeric()
     */
    IsAlphaNumeric(str) {
        for (let i = 0; i < str.length; i++) {
            const code = str.charCodeAt(i);
            if (!(code > 47 && code < 58) && // 0-9
                !(code > 64 && code < 91) && // A-Z
                !(code > 96 && code < 123)) { // a-z
                return false;
            }
        }
        return true;
    }
}
